sql/000005.sql

```diff
@@ -94,11 +104,81 @@ begin
       coalesce(run_at, now()),
       coalesce(max_attempts, 25),
       job_key,
-      coalesce(priority, 0)
+      coalesce(priority, 0),
+      (
+        select jsonb_object_agg(flag, true)
+        from unnest(flags) as item(flag)
+      )
     )
     returning *
     into v_job;
 
   return v_job;
 end;
 $$ language plpgsql volatile;
+
+drop function :GRAPHILE_WORKER_SCHEMA.get_job(text, text[], interval);
+create function :GRAPHILE_WORKER_SCHEMA.get_job(
+  worker_id text,
+  task_identifiers text[] = null,
+  job_expiry interval = interval '4 hours',
+  forbidden_flags text[] = null
+) returns :GRAPHILE_WORKER_SCHEMA.jobs as $$
+declare
+  v_job_id bigint;
+  v_queue_name text;
+  v_row :GRAPHILE_WORKER_SCHEMA.jobs;
+  v_now timestamptz = now();
+begin
+  if worker_id is null or length(worker_id) < 10 then
+    raise exception 'invalid worker id';
+  end if;
+
+  select jobs.queue_name, jobs.id into v_queue_name, v_job_id
+    from :GRAPHILE_WORKER_SCHEMA.jobs
+    where (jobs.locked_at is null or jobs.locked_at < (v_now - job_expiry))
+    and (
+      jobs.queue_name is null
+    or
+      exists (
+        select 1
+        from :GRAPHILE_WORKER_SCHEMA.job_queues
+        where job_queues.queue_name = jobs.queue_name
+        and (job_queues.locked_at is null or job_queues.locked_at < (v_now - job_expiry))
+        for update
+        skip locked
+      )
+    )
+    and run_at <= v_now
+    and attempts < max_attempts
+    and (task_identifiers is null or task_identifier = any(task_identifiers))
+    and (forbidden_flags is null or (flags ?| forbidden_flags) is not true)
```

Comment:
In my limited experience with performance optimizations, nullable arguments in clauses such as `where $1 is null or $1 ...` force the planner not to use the index, but I'm not 100% sure about that, I just remember that this variation worked faster `where $1 is not null and $1 ...`

Reply:
quick test on my own dataset
![image](https://user-images.githubusercontent.com/4533828/91661231-5901fe00-eae3-11ea-858c-f9d3faf6d20a.png)
![image](https://user-images.githubusercontent.com/4533828/91661244-6919dd80-eae3-11ea-8a47-3dd7e298e1d3.png)

Reply:
Thanks for taking the time to analyse this :) The conditions you've displayed are not equivalent. The first would return a result where `phone` is null (desired), the latter will not. We're broadly dealing with the situation where this column would be null in most cases, this is not expected to be an index concern (the field isn't indexed).

The syntax avoiding `OR` for `(phone is null) or (phone like '1111111')` would be `not ((phone is not null) and (phone not like '1111111'))`, and maybe that would be faster if we were indexing this field as it would allow excluding rows where there was an index match; I'm not sure though - it's a pretty broad condition to say `phone not like '1111111'`, I suspect PostgreSQL would opt for a table scan?